web/src/test/java/org/openmrs/web/filter/update/UpdateFilterModelTest.java

```diff
@@ -77,19 +78,20 @@ public void createUpdateFilterModel_shouldrequireAnUpdateAndSetChangesToUnrunDat
 	public void createUpdateFilterModel_shouldRequiredAnUpdateIfChangesAreEmptyButDatabaseUpdaterDoesRequireAnUpdate()
 	        throws Exception {
 		List<OpenMRSChangeSet> changes = new ArrayList<>();
-		
-		when(DatabaseUpdater.getUnrunDatabaseChanges(any(LiquibaseProvider.class))).thenReturn(changes);
-		when(DatabaseUpdater.isLocked()).thenReturn(false);
-		when(DatabaseUpdater.updatesRequired()).thenReturn(true);
-		
+
+		ChangeLogDetective changeLogDetective = new ChangeLogDetective();
+		ChangeLogVersionFinder changeLogVersionFinder = new ChangeLogVersionFinder();
+				
 		model = new UpdateFilterModel(liquibaseProvider);
+		FieldUtils.writeField(model, "updateRequired", true, true);
+		FieldUtils.writeField(model, "changes", changes, true);
+		FieldUtils.writeField(changeLogDetective, "changeLogVersionFinder", changeLogVersionFinder, true);
+		Whitebox.setInternalState(DatabaseUpdater.class, "changeLogDetective", changeLogDetective);
 		
-		assertTrue("should require an update", model.updateRequired);
+		assertTrue(model.updateRequired, "should require an update");
 		assertThat(model.changes, is(empty()));
 		
-		PowerMockito.verifyStatic(DatabaseUpdater.class);
 		DatabaseUpdater.getUnrunDatabaseChanges(liquibaseProvider);
```

Comment:
the error is with this method . i have some difficulty to manipulate the behavior of this method .

Reply:
`java.lang.IllegalStateException: identifying the snapshot version that had been used to initialize the OpenMRS database failed as no candidate change set resulted in zero un-run changes
`

Reply:
the exception is cause by this line 
https://github.com/openmrs/openmrs-core/blob/02ba3aed4e9df2de1261f284f8b4a7cab5eff789/api/src/main/java/org/openmrs/liquibase/ChangeLogDetective.java#L116

i have try to solve the error with this` FieldUtils.writeField(changeLogDetective, "changeLogVersionFinder", changeLogVersionFinder, true);` but it did  not work

Reply:
I am not sure I understand what you want to do. changeLogVersionFinder field is of type ChangeLogVersionFinder so you cannot set it to true since that value is a different type.

Reply:
@WolfSchlegel can you help us out here. So with JUnit 5 PowerMock and thus mocking static methods does not work anymore. We are migrating all test to JUnit 5. This might be the last one we have to migrate. What we can do is use reflection to set static fields. Can you help us out with replacing powermock in this test with reflection so we keep its original intent.

Reply:
Looking into this and apologies for the delay...

`org/openmrs/web/filter/update/UpdateFilterModelTest.java` is a unit test and extending `BaseContextSensitiveTest` makes it an integration test. I would advise against that and would revert that change first.

In the context of https://issues.openmrs.org/browse/TRUNK-4830 I wrote a few new integration tests and found that tests based on `BaseContextSensitiveTest`are not well isolated and understanding the behaviour of individual tests was quite hard. Eventually I created a new base class for integration tests.

@achilep could you please change the test so that it is a unit test again and ping me when this is done?

Thank you,
Wolf

Reply:
sorry @WolfSchlegel for the delay . i restarted every thing back . this is the initial change .

Reply:
Thanks, I have a look

Reply:
Hey there,

please have a look at https://github.com/WolfSchlegel/openmrs-core/commit/6a087975d80b5a592a9209d2ccd8b9c7e611cf6f.

I introduced a wrapper for the (mostly static) `org/openmrs/util/DatabaseUpdater.java` class so that PowerMock is no longer needed for mocking it. This is good news for `org/openmrs/web/filter/update/UpdateFilterModelTest.java`.

However, this comes at a cost as the next question is how to unit test the new wrapper without resorting to PowerMock. I gave JMockit a short try but bumped into issues that are (allegedly) resolved by rearranging the order of dependencies in the pom file (yuck) and it seems a java agent needs to be registered as well for JMockit (yuck-ish).

So I am back at square one looking for a unit test that proves that the all new `org/openmrs/util/DatabaseUpdaterWrapper.java` class plays nicely with the incumbent `org/openmrs/util/DatabaseUpdater.java` class.

To be frank, I am not sure how Java reflections or respective Spring utilities help here. At the end of the day I need to verify that the wrapper calls the (static) methods. For this purpose I introduced some static helper variables and a helper method in `org/openmrs/util/DatabaseUpdater.java`.

Please have a look and let me know what you think. I believe we are facing a trade-off between not testing something and adding test helpers in production code.

Reply:
thank you @WolfSchlegel !

I think your suggestion for adding a wrapper is good. I added a suggestion in the code at the commit you sent.

Our goal with the UpdateModelFilterTest is to test that the filter makes sure updates are made. Your wrapper enables us to do just that. The update logic should be tested else were.

I am ok with not testing the wrapper as its super thin. This tradeoff is also ok for me since it enables us to remove powermock which hopefully forces us to rethink our designs going forward. The proliferation of static methods with side effects in the code base are a sign of bad design for me that circumvent springs dependency injection and harm testability.

Can you please bring in your changes while keeping the UpdateModelFilterTest a JUnit 4 test. We can chat about my suggestion on the code. @achilep  will then do the migration to JUnit 5 once the changes are merged.

Reply:
thanks @WolfSchlegel

Reply:
Thanks for your feedback, @teleivo . 

I removed the hand-rolled verification from the `DatabaseUpdater` class and moved the new wrapper to the same package that contains the `UpdateFilterModel` class. Making the wrapper package-private resulted in flaky Mockito behaviour when running `mvn clean install` so the wrapper remained a public class.

I created this [pull request](https://github.com/achilep/openmrs-core/pull/15) which is based on the current OpenMRS master. If you do not want to get an implicit rebase against master in mid flight, best cherry pick the relevant changes from [here](https://github.com/WolfSchlegel/openmrs-core/commit/0a5f9fe1d198ca2177a50bb44d3ecd5bf6f5c36f).